apps/src/code-studio/components/ExportDialog.jsx

```diff
@@ -333,28 +416,114 @@ class ExportDialog extends React.Component {
     } = await this.publishExpoExport();
 
     if (!expoSnackId) {
-      // We failed to generate a snackId
+      // We failed to generate a snackId, simply return
+      // (publishExpoExport() will have set exportError in state as needed)
       return;
     }
 
     this.setState({generatingApk: true});
     try {
-      const apkUri = await exportApp({
+      const apkBuildId = await exportApp({
         mode: 'expoGenerateApk',
+        md5SavedSources,
         expoSnackId,
         iconUri,
         splashImageUri
       });
+      this.setState({apkBuildId});
+      return this.waitForApkBuild(apkBuildId, expoSnackId);
+    } catch (e) {
       this.setState({
         generatingApk: false,
-        apkError: null,
-        apkUri
+        apkError: 'Failed to create Android app. Please try again later.',
+        apkUri: null,
+        apkBuildId: null
       });
+    }
+  }
+
+  checkForApkBuild(apkBuildId, expoSnackId) {
+    const {exportApp, md5SavedSources} = this.props;
+
+    return exportApp({
+      mode: 'expoCheckApkBuild',
+      md5SavedSources,
+      expoSnackId,
+      apkBuildId
+    });
+  }
+
+  async waitForApkBuild(apkBuildId, expoSnackId) {
+    this.clearWaitTimer();
+
+    try {
+      const apkUri = await this.checkForApkBuild(apkBuildId, expoSnackId);
+      const {generatingApk} = this.state;
+      if (!generatingApk) {
+        // Build was canceled while we were checking on the status
+        return;
+      }
+      if (apkUri) {
+        this.setState({
+          generatingApk: false,
+          apkError: null,
+          apkUri
+        });
+      } else {
+        // Check status again...
+        // NOTE: we don't timeout automatically
+        this.waitTimerId = setTimeout(async () => {
```

Comment:
Why does this need to be an async function?

Reply:
It does not, thanks for catching. Previously, there was an `await` in this function and when that got changed, I never removed the `async`. Will update in a future PR.